packages/remirror__extension-link/src/link-extension.ts

```diff
@@ -739,6 +835,91 @@ export class LinkExtension extends MarkExtension<LinkOptions> {
 
     return autoLinkAllowedTLDs.includes(tld);
   }
+
+  private findURL(input: string): string | undefined {
+    for (const match of findMatches(input, this.options.autoLinkRegex)) {
+      const urlMatch = getMatchString(match);
+
+      if (!this.isValidUrl(urlMatch)) {
+        continue;
+      }
+
+      return urlMatch;
+    }
+
+    return;
+  }
+
+  private getURLEndIndex({
+    url,
+    startIndex = 0,
+    text,
+  }: {
+    url?: string;
+    startIndex?: number;
+    text: string;
+  }) {
+    const newUrl = createNewURL(text);
+
+    // Check for trailing punctuation and adjacent characters
+    if (newUrl) {
+      const endsWithPunctuation = (input: string) =>
+        this.options.adjacentPunctuations.includes(input.slice(-1));
+
+      const adjacentCharCount = getTrailingCharCount({
+        input: text,
+        url,
+      });
+
+      const { pathname, search } = newUrl;
+
+      let decoded;
+
+      try {
+        decoded = decodeURI(pathname.slice(1) + search);
+      } catch {
+        decoded = pathname.slice(1) + search;
+      }
+
+      if (decoded && endsWithPunctuation(decoded)) {
+        const index = -1;
+        const balancedIndex = !isBalanced(decoded) ? getBalancedIndex(decoded, index) : 0;
+        const trailingPunctuationIndex = SENTENCE_PUNCTUATIONS.includes(decoded.slice(-1))
+          ? getTrailingPunctuationIndex(decoded, index)
+          : 0;
+
+        if (balancedIndex < 0) {
+          const balanced = decoded.slice(0, balancedIndex);
+
+          if (SENTENCE_PUNCTUATIONS.includes(balanced.slice(-1))) {
+            return getTrailingPunctuationIndex(balanced, index) + balancedIndex;
+          }
+
+          return balancedIndex;
+        }
+
+        if (trailingPunctuationIndex < 0) {
+          return trailingPunctuationIndex;
+        }
+
+        return;
+      }
+
+      if (adjacentCharCount && !this.isValidUrl(text.slice(startIndex))) {
+        return (
+          adjacentCharCount +
+          // Adjust index when switching to longer TLD e.g '.co'  to '.com'
+          (this.isValidUrl(text.slice(startIndex, adjacentCharCount + 1)) ? 1 : 0)
+        );
+      }
+
+      if (endsWithPunctuation(text)) {
+        return -1;
```

Comment:
Think we're assuming here that is ends with a single punctuation mark? Part of the reason why I think `endsWithPunctation` should return the string instead?

Reply:
Think this is resolved, otherwise let me know, please.